Interrogator.xUnit/NotConcurrentAttribute.cs

```diff
@@ -0,0 +1,183 @@
+				if (methods.Length == 0)
+					continue;
+
+				foreach (var method in methods)
+				{
+					var attributes = method.GetCustomAttributes(true).OfType<NotConcurrentAttribute>();
+					foreach (var attribute in attributes)
+					{
+						var key = (type, attribute.GroupName);
+						if (!_methodInfoDictionary.ContainsKey(key))
+							_methodInfoDictionary.Add(key, new List<MethodInfo>());
+						_methodInfoDictionary[key].Add(method);
+					}
+				}
+
+			}
+		}
+
+		public enum ConcurrencyScope
+		{
+			Class,
+			ClassHierarchy,
+			Namespace,
+			Assembly
+		}
+
+		public string GroupName { get; }
+		public ConcurrencyScope Scope { get; }
+
+		public NotConcurrentAttribute() : this(ConcurrencyScope.Class, String.Empty) { }
+		public NotConcurrentAttribute(string groupName) : this(groupName, ConcurrencyScope.Class) { }
+
+		public NotConcurrentAttribute(ConcurrencyScope scope) : this(scope, String.Empty) { }
+
+		public NotConcurrentAttribute(string groupName, ConcurrencyScope scope) : this(scope, groupName)
+		{
+			if (String.IsNullOrWhiteSpace(groupName))
+				throw new InvalidGroupNameException(groupName);
+		}
+
+		private NotConcurrentAttribute(ConcurrencyScope scope, string groupName) : base(null)
+		{
+			GroupName = $"{groupName}_{scope}"; // Adding scope to group name ensures that different scoped attributes don't collide
+			Scope = scope;
+		}
+
+		internal override Result<Option<MethodInfo>, string> TryGetMethod(Type containingType, MemberInfo member, MethodInfo[] testMethods)
+		{
+			var result = (Type ?? containingType)
+				.TryGetMethod(member.Name, ParameterTypes)
+				.Select(currentMethod => GetPreviousInChain(currentMethod, containingType, Scope, GroupName, testMethods));
+			return result;
+		}
+
+		private static Option<MethodInfo> GetPreviousInChain(MethodInfo currentMethod, Type type, ConcurrencyScope scope, string groupName, MethodInfo[] executingMethods)
+		{
+			var methodSet = new HashSet<MemberInfo>(executingMethods);
+			var methodsToExecute = GetSameGroupMethods(scope, type, groupName).Where(methodSet.Contains).ToArray();
+
+			return GetPreviousMethod(currentMethod, methodsToExecute);
+		}
+
+		private static MethodInfo[] GetSameGroupMethods(ConcurrencyScope scope, Type type, string groupName)
+		{
+			IEnumerable<Type> types;
+			switch (scope)
+			{
+				case ConcurrencyScope.Class:
+					types = new[] { type };
+					break;
+				case ConcurrencyScope.Assembly:
+					types = GetTypesInAssembly(type);
+					break;
+				case ConcurrencyScope.Namespace:
+					types = GetTypesInNamespace(type);
+					break;
+				case ConcurrencyScope.ClassHierarchy:
+					types = GetTypesInClassHierarchy(type);
+					break;
+				default:
+					types = Array.Empty<Type>();
+					break;
+			}
+
+			var list = new List<MethodInfo>();
+			foreach (var t in types)
+			{
+				var key = (t, groupName);
+				if (_methodInfoDictionary.ContainsKey(key))
+					list.AddRange(_methodInfoDictionary[key]);
+			}
+
+			return list.ToArray();
```

Comment:
I think you should be ordering the types here and not in GetMethodsInTypes()

Reply:
I think the final list needs to be sorted, right? The methods for each type are already sorted, so you probably just need to sort the types up in the foreach.